ice-adapter/src/main/java/com/faforever/iceadapter/IceAdapter.java

```diff
@@ -113,42 +126,45 @@ public static void onDisconnectFromPeer(int remotePlayerId) {
         });
     }
 
-    private static synchronized void createGameSession() {
-        if (GAME_SESSION != null) {
-            GAME_SESSION.close();
-            GAME_SESSION = null;
-        }
+    private static void createGameSession() {
+        LockUtil.executeWithLock(lockGameSession, () -> {
+            if (GAME_SESSION != null) {
+                GAME_SESSION.close();
+                GAME_SESSION = null;
+            }
 
-        GAME_SESSION = new GameSession();
+            GAME_SESSION = new GameSession();
+        });
     }
 
     /**
      * Triggered by losing gpgnet connection to FA.
      * Closes the active Game/ICE session
      */
-    public static synchronized void onFAShutdown() {
-        if (GAME_SESSION != null) {
-            log.info("FA SHUTDOWN, closing everything");
-            GAME_SESSION.close();
-            GAME_SESSION = null;
-            // Do not put code outside of this if clause, else it will be executed multiple times
-        }
+    public static void onFAShutdown() {
+        LockUtil.executeWithLock(lockGameSession, () -> {
+            if (GAME_SESSION != null) {
+                log.info("FA SHUTDOWN, closing everything");
+                GAME_SESSION.close();
+                GAME_SESSION = null;
+                // Do not put code outside of this if clause, else it will be executed multiple times
+            }
+        });
     }
 
     /**
      * Stop the ICE adapter
      */
-    public static void close() {
+    public static void close(int status) {
         log.info("close() - stopping the adapter");
 
-        Executor.executeDelayed(500, () -> System.exit(0));
-
         onFAShutdown(); // will close gameSession aswell
         GPGNetServer.close();
         RPCService.close();
         TrayIcon.close();
+        INSTANCE.close();
```

Comment:
I'm not sure what the point here is, if we exit right away. The code does not wait for the shutdownNow to take place after 1 s.

Reply:
Changed it, now she application waits a bit before it finishes.